File: SkylinkSample/SkylinkSample/MessagesViewController.m
//
//  MessagesViewController.m
//  Skylink_Examples
//
//  Created by Temasys on 04/01/2016.
//  Copyright © 2016 Temasys. All rights reserved.
//

#import "MessagesViewController.h"
#import "Constant.h"
#import "SAMessage.h"

//#define ROOM_NAME [[NSUserDefaults standardUserDefaults] objectForKey:@"ROOMNAME_MESSAGES"]


@interface MessagesViewController ()<SKYLINKConnectionMessagesDelegate, UIPickerViewDelegate, UIPickerViewDataSource, UITextFieldDelegate>
// IBOutlets
@property (weak, nonatomic) IBOutlet UITextField *messageTextField;
@property (weak, nonatomic) IBOutlet UITableView *tableView;
@property (weak, nonatomic) IBOutlet UISwitch *isPublicSwitch;
@property (weak, nonatomic) IBOutlet UISegmentedControl *messageTypeSegmentControl;
@property (weak, nonatomic) IBOutlet UITextField *nicknameTextField;
@property (weak, nonatomic) IBOutlet UIButton *peersButton;
@property (weak, nonatomic) IBOutlet UIButton *sendButton;
@property (weak, nonatomic) IBOutlet UIActivityIndicatorView *activityIndicator;
@property (weak, nonatomic) IBOutlet UITextField *encryptKeyTextField;
@property (weak, nonatomic) IBOutlet UIView *pickerViewContainer;
@property (weak, nonatomic) IBOutlet UIPickerView *pickerView;
@property (weak, nonatomic) IBOutlet UISwitch *persistSwitch;

// Properties
@property (strong, nonatomic) NSMutableArray *messages;
@property (strong, nonatomic) NSMutableDictionary *peers;

@end


@implementation MessagesViewController{
    NSArray *_encryptSecretIds;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    
    self.title = @"Messages";
    self.messages = [NSMutableArray new];
    self.peers = [NSMutableDictionary new];
    [self updatePeersButtonTitle];
    [self loadStoredMessage];
    // Creating configuration
    SKYLINKConnectionConfig *config = [SKYLINKConnectionConfig new];
    [config setAudioVideoSendConfig:AudioVideoConfig_NO_AUDIO_NO_VIDEO];
    [config setAudioVideoReceiveConfig:AudioVideoConfig_NO_AUDIO_NO_VIDEO];
    config.hasP2PMessaging = YES;
    // Creating SKYLINKConnection
    _skylinkConnection = [[SKYLINKConnection alloc] initWithConfig:config callback:nil];
    _skylinkConnection.lifeCycleDelegate = self;
    _skylinkConnection.messagesDelegate = self;
    _skylinkConnection.remotePeerDelegate = self;
    _skylinkConnection.encryptSecrets = SAConstants.shared.ENCRYPTION_SECRETS;
    // Connecting to a room
    [self joinRoom];
    _encryptSecretIds = [@[@"No Key"] arrayByAddingObjectsFromArray:[SAConstants.shared.ENCRYPTION_SECRETS.allKeys sortedArrayUsingSelector:@selector(compare:)]];
    _encryptKeyTextField.text = _encryptSecretIds.firstObject;
}

#pragma mark - SKYLINKConnectionLifeCycleDelegate
- (void)connectionDidConnectToRoomSuccessful:(SKYLINKConnection *)connection
{
    MyLog(@"Inside %s", __FUNCTION__);
    dispatch_async(dispatch_get_main_queue(), ^{
        self.messageTextField.enabled = YES;
        self.messageTextField.hidden = NO;
        self.nicknameTextField.enabled = YES;
        self.nicknameTextField.hidden = NO;
        self.sendButton.enabled = YES;
        self.sendButton.hidden = NO;
        [self.messageTextField becomeFirstResponder];
        [self.activityIndicator stopAnimating];
    });
}

- (void)connection:(SKYLINKConnection *)connection didConnectToRoomFailed:(NSString *)errorMessage
{
    [UIAlertController showAlertWithAutoDisappearTitle:@"Connection failed" message:errorMessage duration:3 onViewController:self];
    [self.navigationController popViewControllerAnimated:YES];
}

- (void)connection:(SKYLINKConnection *)connection didDisconnectFromRoomWithSkylinkEvent:(NSDictionary *)skylinkEvent contextDescription:(NSString *)contextDescription
{
//    [UIAlertController showAlertWithAutoDisappearTitle:@"Disconnected" message:contextDescription duration:3 onViewController:self];
//    [self.navigationController popViewControllerAnimated:YES];
}


#pragma mark - SKYLINKConnectionMessagesDelegate
- (void)connection:(SKYLINKConnection *)connection didReceiveServerMessage:(id)message isPublic:(BOOL)isPublic timeStamp:(long long)timeStamp remotePeerId:(NSString *)remotePeerId{
    NSLog(@"SIG message");
    if ([message isKindOfClass:[NSString class]]) {
        SAMessage *receivedMsg = [[SAMessage alloc] initWithData:message timeStamp:timeStamp sender:[self getUserNameFrom:remotePeerId] target:(isPublic ? nil : _skylinkConnection.localPeerId) type:SAMessageTypeP2P];
        [_messages addObject:receivedMsg];
        [_tableView reloadData];
    }
}
- (void)connection:(SKYLINKConnection *)connection didReceiveP2PMessage:(id)message isPublic:(BOOL)isPublic timeStamp:(long long)timeStamp remotePeerId:(NSString *)remotePeerId{
    NSLog(@"P2P message");
    if ([message isKindOfClass:[NSString class]]) {
        SAMessage *receivedMsg = [[SAMessage alloc] initWithData:message timeStamp:timeStamp sender:[self getUserNameFrom:remotePeerId] target:(isPublic ? nil : _skylinkConnection.localPeerId) type:SAMessageTypeP2P];
        [_messages addObject:receivedMsg];
        [_tableView reloadData];
    }
}

- (void)connection:(SKYLINKConnection *)connection didReceiveP2PMessage:(id)message isPublic:(BOOL)isPublic remotePeerId:(NSString *)remotePeerId
{
    [self.messages insertObject:@{@"message" : message, @"isPublic" : @(isPublic), @"peerId" : remotePeerId, @"type" : @"P2P"} atIndex:0];
    [self.tableView reloadSections:[NSIndexSet indexSetWithIndex:0] withRowAnimation:UITableViewRowAnimationFade];
}

#pragma mark - SKYLINKConnectionRemotePeerDelegate
- (void)connection:(SKYLINKConnection *)connection didConnectWithRemotePeer:(NSString *)remotePeerId userInfo:(id)userInfo hasDataChannel:(BOOL)hasDataChannel
{
    NSString *displayNickname = (userInfo && [userInfo isKindOfClass:[NSDictionary class]] && userInfo[@"nickname"]) ? userInfo[@"nickname"] : [NSString stringWithFormat:@"ID: %@", remotePeerId];
    [self.peers addEntriesFromDictionary:@{remotePeerId:displayNickname}];
    [self updatePeersButtonTitle];
    [self.tableView reloadData];
    [self.activityIndicator stopAnimating];
}

- (void)connection:(SKYLINKConnection *)connection didReceiveRemotePeerLeaveRoom:(NSString *)remotePeerId userInfo:(id)userInfo skylinkInfo:(NSDictionary *)skylinkInfo
{
    MyLog(@"Peer with ID %@ left with skylinkInfo: %@", remotePeerId, skylinkInfo);
    [self.peers removeObjectForKey:remotePeerId];
    [self updatePeersButtonTitle];
}
#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.messages.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"messageCell"];
    
    SAMessage *message = [_messages objectAtIndex: _messages.count - indexPath.row - 1];
    
    cell.textLabel.text = [NSString stringWithFormat:@"%@~~~%@", [message timeStampString], message.data];
    cell.detailTextLabel.text = [NSString stringWithFormat:@"From %@ via %@ • %@", message.sender, message.typeToString, [message isPublicString]];
    return cell;
}

#pragma mark - Table view delegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    SAMessage *message = [self.messages objectAtIndex:_messages.count - indexPath.row - 1];
    NSString *messageDetails = [NSString stringWithFormat:@"Message:\n%@\n\nFrom :\n%@\n\n%@", message.data, ([message.sender isEqualToString:USER_NAME] ? @"me" : message.sender), [message isPublicString]];
    showAlert(@"Message Detail", messageDetails);
}

#pragma mark - IBActions

- (IBAction)sendTap:(UIButton *)sender {
    _skylinkConnection.messagePersist = _persistSwitch.isOn;
    NSString *message = _messageTextField.text;
    if (_peers.count<=0) {
        showAlert(@"No peer connected", @"\nYou can't define a private recipient since there is no peer connected.");
        return;
    }
    
    if (![message isNotEmpty]) {
        showAlert(@"Empty Message", @"\nType the message to be sent.");
        return;
    }
    
    if (_isPublicSwitch.isOn) {
        //send public
        [self sendMessage:message forPeerId:nil];
    }else{
        //send private
        UIAlertController *_alert = [UIAlertController alertControllerWithTitle:@"Choose a private recipient." message:@"\nYou're about to send a private message\nWho do you want to send it to ?" preferredStyle:UIAlertControllerStyleAlert];
        UIAlertAction *_cancelAction = [UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleDefault handler:nil];
        __weak __typeof(self)weakSelf = self;
        for (NSString *peerDicKey in _peers.allKeys) {
            UIAlertAction *_peerAction = [UIAlertAction actionWithTitle:_peers[peerDicKey] style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
                [weakSelf sendMessage:message forPeerId:peerDicKey];
//                [weakSelf alert:self->_peers[peerDicKey] message:message];
            }];
            [_alert addAction:_peerAction];
        }
        [_alert addAction:_cancelAction];
        [self presentViewController:_alert animated:YES completion:nil];
    }
}

- (IBAction)dismissKeyboardTap:(UIButton *)sender {
    [self hideKeyboardIfNeeded];
}

- (IBAction)peersTap:(UIButton *)sender {
    [UIAlertController showAlertWithAutoDisappearTitle:sender.titleLabel.text message:[self.peers description] duration:3 onViewController:self];
}
- (IBAction)doneEncryptSecret:(UIButton *)sender{
    [self.pickerViewContainer setHidden:YES];
}


#pragma mark - Utils
-(void)alert:(NSString *)title message:(NSString*)message{
    showAlert(title, message);
    [self sendMessage:message forPeerId:[title stringByReplacingOccurrencesOfString:@"ID: " withString:@""]];
}

- (void)sendMessage:(NSString *)message forPeerId:(NSString *)peerId { // nil peerId means public message
    void (^processResponse)(NSError *, SAMessageType) = ^(NSError *error, SAMessageType type){
        if (error) {
            showAlert([NSString stringWithFormat:@"ERROR: %ld", (long)error.code], error.localizedDescription);
        }else{
            SAMessage *msg = [[SAMessage alloc] initWithData:message timeStamp:[[NSDate date] toTimeStamp] sender:USER_NAME target:peerId type:type];
            [self->_messages addObject:msg];
            self.messageTextField.text = @"";
            [self.tableView reloadData];
            showAlert(message, peerId ? peerId : @"All");
        }
    };
    if (_messageTypeSegmentControl.selectedSegmentIndex) {
        [_skylinkConnection sendServerMessage:message toRemotePeerId:peerId callback:^(NSError * _Nullable error) {
            processResponse(error, SAMessageTypeSignaling);
        }];
    }else{
        [_skylinkConnection sendP2PMessage:message toRemotePeerId:peerId callback:^(NSError * _Nullable error) {
            processResponse(error, SAMessageTypeP2P);
        }];
    }
    
}

- (void)updateNickname {
    if (self.nicknameTextField.text.length > 0) [_skylinkConnection sendLocalUserData:@{@"nickname" : self.nicknameTextField.text} callback:^(NSError * _Nullable error) {
        if (error) [UIAlertController showAlertWithAutoDisappearTitle:@"Error" message:error.localizedDescription duration:3 onViewController:self];
    }];
    else [UIAlertController showAlertWithAutoDisappearTitle:@"Empty nickname" message:@"\nType the nickname to set." duration:3 onViewController:self];
}

- (void)updatePeersButtonTitle {
    NSUInteger peersCount = self.peers.count;
    if (peersCount == 0) [self.peersButton setTitle:@"No peer" forState:UIControlStateNormal];
    else [self.peersButton setTitle:[NSString stringWithFormat:@"%lu peer%@", (unsigned long)peersCount, (peersCount > 1) ? @"s" : @""] forState:UIControlStateNormal];
}

- (void)hideKeyboardIfNeeded {
    [self.messageTextField resignFirstResponder];
    [self.nicknameTextField resignFirstResponder];
    [self.encryptKeyTextField resignFirstResponder];
}
- (void)loadStoredMessage{
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(5 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        [self->_skylinkConnection getStoredMessages:^(NSArray * _Nullable storedMessages, NSDictionary * _Nullable errors) {
            if (!self.view.window) {
                return;
            }
            if (errors) {
                showAlert(@"Error map", errors.description);
            }
            for (NSDictionary *item in storedMessages) {
                if ([item isKindOfClass:[NSDictionary class]]) {
                    SAMessage *message = [[SAMessage alloc] initWithData:item[@"data"] timeStamp:[item[@"timeStamp"] longLongValue] sender:[self getUserNameFrom:item[@"peerId"]] target:nil type:SAMessageTypeSignaling];
                    [self.messages addObject:message];
                }
            }
            [self.tableView reloadData];
        }];
    });
}
- (NSString *)getUserNameFrom:(NSString *)peerId{
    NSDictionary *userInfo = [_skylinkConnection getUserInfo:peerId];
    if (userInfo) {
        return userInfo[@"userData"];
    }
    return peerId;
}

#pragma mark - UITextField delegate

- (BOOL)textFieldShouldReturn:(UITextField *)textField {
    if ([textField isEqual:self.nicknameTextField]) [self updateNickname];
    [self hideKeyboardIfNeeded];
    return YES;
}
- (BOOL)textFieldShouldBeginEditing:(UITextField *)textField{
    if (textField == self.encryptKeyTextField) {
        [self hideKeyboardIfNeeded];
        [_pickerViewContainer setHidden:NO];
        return NO;
    }
    return YES;
}
#pragma mark - PICKER VIEW
- (NSInteger)numberOfComponentsInPickerView:(UIPickerView *)pickerView{
    return 1;
}
- (NSInteger)pickerView:(UIPickerView *)pickerView numberOfRowsInComponent:(NSInteger)component{
    return _encryptSecretIds.count;
}

- (NSString *)pickerView:(UIPickerView *)pickerView titleForRow:(NSInteger)row forComponent:(NSInteger)component{
    return _encryptSecretIds[row];
}

- (void)pickerView:(UIPickerView *)pickerView didSelectRow:(NSInteger)row inComponent:(NSInteger)component{
    NSString *selectedEncryptSecret = (row == 0) ? nil : _encryptSecretIds[row];
    _skylinkConnection.selectedSecretId = selectedEncryptSecret;
    _encryptKeyTextField.text = _encryptSecretIds[row];
}
@end

File: Swift_version/sampleapp/Sample_App/Sample_App/MessagesViewController.swift
//
//  MessagesViewController.swift
//  Sample_App
//
//  Created by Phyo Pwint  on 12/4/16.
//  Copyright © 2016 Temasys . All rights reserved.
//

import UIKit
import AVFoundation


class MessagesViewController: UIViewController,SKYLINKConnectionLifeCycleDelegate, SKYLINKConnectionMessagesDelegate, SKYLINKConnectionRemotePeerDelegate, UITextFieldDelegate  {
    
    //Declaring the Elements
    
    let ROOM_NAME = "MESSAGES-ROOM"
    var messages : NSMutableArray!
    var peers : NSMutableDictionary!
    var skylinkConnection: SKYLINKConnection!
    var skylinkApiKey: String!
    var skylinkApiSecret: NSString!
    var topView: UIView!
    
    @IBOutlet weak var activityIndicator: UIActivityIndicatorView!
    @IBOutlet weak var sendButton: UIButton!
    @IBOutlet weak var peersButton: UIButton!
    @IBOutlet weak var nicknameTextField: UITextField!
    @IBOutlet weak var messageTextField: UITextField!
    @IBOutlet weak var tableView: UITableView!
    @IBOutlet weak var messageTypeSegmentControl: UISegmentedControl!
    @IBOutlet weak var isPublicSwitch: UISwitch!
    
    
    override func viewDidLoad() {
        super.viewDidLoad()
        nicknameTextField.delegate = self
        messageTextField.delegate = self
        NSLog("SKYLINKConnection version = %@", SKYLINKConnection.getSkylinkVersion())
        self.title = "Messages"
        self.navigationItem.leftBarButtonItem = UIBarButtonItem(image: UIImage(named: "Cancel.png"), style: .Plain, target: self, action: #selector(MessagesViewController.disConnect))
        let infoButton: UIButton = UIButton(type: .InfoLight)
        infoButton.addTarget(self, action: #selector(MessagesViewController.showInfo), forControlEvents: .TouchUpInside)
        self.navigationItem.rightBarButtonItem = UIBarButtonItem(customView: infoButton)
        self.messages = []
        self.peers = [:]
        self.updatePeersButtonTitle()
        //Creating configuration
        let config:SKYLINKConnectionConfig = SKYLINKConnectionConfig()
        config.video = false
        config.audio = false
        config.fileTransfer = false
        config.dataChannel = true // for data chanel messages
        
        // Creating SKYLINKConnection
        self.skylinkConnection = SKYLINKConnection(config: config, appKey: self.skylinkApiKey! as String)
        self.skylinkConnection.lifeCycleDelegate = self
        self.skylinkConnection.messagesDelegate = self
        self.skylinkConnection.remotePeerDelegate = self
        
        // Connecting to a room
        SKYLINKConnection.setVerbose(true)
        self.skylinkConnection.connectToRoomWithSecret(self.skylinkApiSecret! as String, roomName: ROOM_NAME, userInfo: nil)
        
    }
    
    func disConnect() {
        if (self.skylinkConnection != nil) {
            self.skylinkConnection.disconnect({() -> Void in
                self.navigationController!.popViewControllerAnimated(true)
            })
        }
    }
    
    func showInfo() {
        let msgTitle: String = "Infos"
        let msg: String = "\nRoom name:\n\(ROOM_NAME)\n\nLocal ID:\n\(self.skylinkConnection.myPeerId)\n\nKey: •••••\(self.skylinkApiKey.substringFromIndex(self.skylinkApiKey.startIndex.advancedBy(self.skylinkApiKey.characters.count-7)))\n\nSkylink version \(SKYLINKConnection.getSkylinkVersion())"
        AlertMessage(msgTitle, msg:msg)
    }
    
    func AlertMessage(msg_title: String, msg:String) {
        let alertController = UIAlertController(title: msg_title , message: msg, preferredStyle: .Alert)
        let OKAction = UIAlertAction(title: "OK", style: .Default) { (action) in }
        alertController.addAction(OKAction)
        self.presentViewController(alertController, animated: true, completion: nil)
    }
    
    // MARK: SKYLINKConnectionLifeCycleDelegate
    func connection(connection: SKYLINKConnection, didConnectWithMessage errorMessage: String, success isSuccess: Bool) {
        if isSuccess {
            NSLog("Inside %@", #function)
            
            dispatch_async(dispatch_get_main_queue(), {() -> Void in
                self.messageTextField.enabled = true
                self.messageTextField.hidden = false
                self.nicknameTextField.enabled = true
                self.nicknameTextField.hidden = false
                self.sendButton.enabled = true
                self.sendButton.hidden = false
                self.messageTextField.becomeFirstResponder()
            })
            
        }
        else {
            let msgTitle: String = "Connection failed"
            let msg: String = errorMessage
            AlertMessage(msgTitle, msg:msg)
            self.disConnect()
        }
        dispatch_async(dispatch_get_main_queue(), {() -> Void in
            self.activityIndicator.stopAnimating()
        })
    }
    
    func connection(connection: SKYLINKConnection, didDisconnectWithMessage errorMessage: String) {
        let msgTitle: String = "Disconnected"
        let alertController = UIAlertController(title: msgTitle , message: errorMessage, preferredStyle: .Alert)
        let OKAction = UIAlertAction(title: "OK", style: .Default) { (action) in }
        alertController.addAction(OKAction)
        self.presentViewController(alertController, animated: true, completion: {
            self.disConnect()
        })
    }
    
    // MARK: SKYLINKConnectionRemotePeerDelegate
    func connection(connection: SKYLINKConnection, didJoinPeer userInfo: AnyObject, mediaProperties pmProperties: SKYLINKPeerMediaProperties, peerId: String) {
        let displayNickName: String = (userInfo.isKindOfClass(NSDictionary) &&
            userInfo["nickname"] != nil) ? userInfo["nickname"] as! String : "ID: \(peerId)"
        self.peers.addEntriesFromDictionary([peerId:displayNickName])
        self.updatePeersButtonTitle()
    }
    
    func connection(connection: SKYLINKConnection, didLeavePeerWithMessage errorMessage: String, peerId: String) {
        self.peers.removeObjectForKey(peerId)
        self.updatePeersButtonTitle()
    }
    
    func connection(connection: SKYLINKConnection, didReceiveUserInfo userInfo: AnyObject, peerId: String) {
        self.peers.removeObjectForKey(peerId)
        let displayNickname: String = ((userInfo["nickname"]) != nil) ? userInfo["nickname"] as! String : "ID: \(peerId)"
        self.peers.addEntriesFromDictionary([peerId:displayNickname])
        self.updatePeersButtonTitle()
        self.tableView.reloadData()
    }
    
    // MARK: SKYLINKConnectionMessagesDelegate
    
    func connection(connection: SKYLINKConnection!, didReceiveCustomMessage message: AnyObject, `public` isPublic: Bool, peerId: String) {
        self.messages.insertObject(
            [   "message": message,
                "isPublic": isPublic,
                "peerId":peerId,
                "type":"signaling server"],
            atIndex : 0)
        self.tableView.reloadSections(NSIndexSet(index: 0), withRowAnimation: .Fade)
    }
    
    func connection(connection: SKYLINKConnection!, didReceiveDCMessage message: AnyObject, `public` isPublic: Bool, peerId: String!) {
        self.messages.insertObject(
            [   "message": message,
                "isPublic": isPublic,
                "peerId":peerId,
                "type":"P2P"],
            atIndex : 0)
        self.tableView.reloadSections(NSIndexSet(index: 0), withRowAnimation: .Fade)
        
    }
    
    func connection(connection: SKYLINKConnection!, didReceiveBinaryData data: NSData, peerId: String) {
        let maybeString: AnyObject = String(data: data, encoding: NSUTF8StringEncoding)!
        
        self.messages.insertObject(["message": ((maybeString is NSString)) ? (String(maybeString)) : "Binary data of length \(UInt(data.length))", "isPublic": false, "peerId": peerId, "type": "binary data"], atIndex: 0)
        
        self.tableView.reloadSections(NSIndexSet(index: 0), withRowAnimation: .Fade)
        
    }
    
    // MARK:  - Table view data source
    func numberOfSectionsInTableView(tableView: UITableView) -> Int {
        return 1
    }
    
    func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.messages.count
    }
    
    func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        let cell: UITableViewCell = tableView.dequeueReusableCellWithIdentifier("messageCell")!
        var message: [NSObject : AnyObject] = (self.messages.objectAtIndex(indexPath.row)) as! [NSObject : AnyObject]
        cell.textLabel!.text = message["message"] as? String
        let equalStr: String = self.skylinkConnection.myPeerId
        
        if((message["peerId"] as! String) == equalStr) {
            cell.detailTextLabel!.text = (message["isPublic"] as! Bool) ? "Sent to all" : "Sent privately"
            cell.backgroundColor = UIColor(red: 0.71, green: 1, blue: 0.5, alpha: 1)
        }
        else {
            cell.detailTextLabel!.text = String(format:"From \(peers[message["peerId"] as! String]!) via \(message["type"] as! String) • \(message["isPublic"] as! Bool ? "Public" : "Private")")
            cell.backgroundColor = UIColor.whiteColor()
        }
        return cell
    }
    
    // MARK: Table view delegate
    
    func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
        tableView.deselectRowAtIndexPath(indexPath, animated: true)
        var message: [NSObject : AnyObject] = self.messages.objectAtIndex(indexPath.row) as! [NSObject : AnyObject]
        let messageDetails: String = String(format:"Message \n %@ \n\n From: \n %@ \n\n %@", message["message"] as! String, (message["peerId"] as! String) == (self.skylinkConnection.myPeerId as String) ? "me" : (message["peerId"] as! String) ,message["isPublic"] as! Bool ? "Public" : "Private")
        let msgTitle: String = "Message Details"
        let msg: String = messageDetails
        AlertMessage(msgTitle, msg: msg)
    }
    
    // MARK: - Utils
    func processMessage() {
        if(self.isPublicSwitch.on && self.messageTypeSegmentControl.selectedSegmentIndex == 2) {
            let msgTitle: String = "Binary data is private."
            let msg: String = "\nTo send your message as binary data, uncheck the \"Public\" UISwitch."
            AlertMessage(msgTitle, msg: msg)
            self.hideKeyboardIfNeeded()
        }
        else if self.messageTextField.text!.characters.count > 0 {
            let message: String = self.messageTextField.text!
            if !(self.isPublicSwitch.on) {
                if self.peers.count != 0 {
                    let alert = UIAlertController(title: "Choose a private recipient.", message: "\nYou're about to send a private message\nWho do you want to send it to ?", preferredStyle: .Alert)
                    
                    let noAction = UIAlertAction(title: "Cancel", style: .Default) { (action) -> Void in
                    }
                    for peerDicKey in self.peers.allKeys as! [String] {
                        let yesAction = UIAlertAction(title: self.peers[peerDicKey] as? String, style: .Default) { (action) -> Void in
                            self.sendMessage(message, forPeerId: peerDicKey)
                        }
                        alert.addAction(yesAction)
                    }
                    alert.addAction(noAction)
                    self.presentViewController(alert, animated: true, completion: nil)
                }
                else {
                    let msgTitle: String = "No peer connected."
                    let msg: String = "\nYou can't define a private recipient since there is no peer connected."
                    AlertMessage(msgTitle, msg: msg)
                }
                
                
            }
            else {
                
                self.sendMessage(message, forPeerId: nil)
            }
        }
        else {
            let msgTitle: String = "Empty message"
            let msg: String = "\nType the message to be sent."
            AlertMessage(msgTitle, msg: msg)
            
        }
        
    }
    
    func sendMessage(message: String, forPeerId peerId: String?) {
        var showSentMessage: Bool = true
        switch(self.messageTypeSegmentControl.selectedSegmentIndex){
        case 0 :
            self.skylinkConnection.sendDCMessage(message, peerId: peerId)
            NSLog("Finish DCMessage")
            break;
        case 1 :
            self.skylinkConnection.sendCustomMessage(message, peerId: peerId)
            break;
        case 2 :
            do {
                if(peerId != nil) {
                    try self.skylinkConnection.sendBinaryData(message.dataUsingEncoding(NSUTF8StringEncoding), peerId: peerId)
                }
            } catch let e as NSError {
                var message: String = "\(e)"
                if(e.localizedDescription == "Sending binary data in a MCU connection is not supported") {
                    message = message.stringByAppendingString("MCU can be enabled/disabled in Key configuration on the developer portal: http://developer.temasys.com.sg/")
                }
                let msgTitle: String = "Exeption when sending binary data"
                let msg: String = message
                AlertMessage(msgTitle, msg: msg)
                showSentMessage = false
            }
            break;
            
        default : break;
            
        }
        
        if(showSentMessage) {
            self.messageTextField.text = ""
            self.messages.insertObject(
                ["message": message,
                    "isPublic": NSNumber(bool: (peerId == nil)),
                    "peerId": self.skylinkConnection.myPeerId],
                atIndex: 0)
            self.tableView.reloadSections(NSIndexSet(index: 0), withRowAnimation: UITableViewRowAnimation .Fade)
        }
        else {
            self.hideKeyboardIfNeeded()
        }
    }
    
    
    func updatePeersButtonTitle() {
        let peersCount: Int = self.peers.count
        if(peersCount == 0) {
            self.peersButton.setTitle("No Peer", forState: .Normal)
        }
        else {
            self.peersButton.setTitle("\(peersCount) peer" + (peersCount > 1 ? "s " : ""), forState: .Normal)
        }
    }
    
    func updateNickname() {
        if (self.nicknameTextField.text!.characters.count > 0) {
            self.skylinkConnection.sendUserInfo(["nickname" : self.nicknameTextField.text!])
        }
        else {
            let msgTitle: String = "Empty nickname"
            let msg: String = "\nType the nickname to set."
            AlertMessage(msgTitle, msg: msg)
        }
    }
    
    func hideKeyboardIfNeeded() {
        self.messageTextField.resignFirstResponder()
        self.nicknameTextField.resignFirstResponder()
    }
    
    // MARK: - UITextField delegate
    func textFieldShouldReturn(textField: UITextField) -> Bool {
        if textField.isEqual(self.nicknameTextField) {
            self.updateNickname()
        }
        else if textField.isEqual(self.messageTextField) {
            self.processMessage()
        }
        self.hideKeyboardIfNeeded()
        return true
    }
    
    // MARK: IBFuction
    @IBAction func sendTap(sender: UIButton) {
        self.processMessage()
    }
    
    @IBAction func dismissKeyboardTap(sender: UIButton) {
        self.hideKeyboardIfNeeded()
    }
    
    @IBAction func peersTap(sender: UIButton) {
        let msgTitle:String = sender.titleLabel!.text!
        let msg:String = self.peers.description
        AlertMessage(msgTitle, msg: msg)
    }
}
